### ISSUE TITLE
network namespaces not being cleaned up on cluster-store bounce

### ISSUE THREAD

**Issue by mgoldsborough** [docker/libnetwork#793]
Hi,

I'm messing around with multi-host networking and hit a few snags.  It seems that the networking namespaces are not cleaned up if the cluster-store goes down while containers are running.  I get the following:

```
node3:~# docker run -it  --rm --name container3  --net=RED busybox
Error response from daemon: Cannot start container e5d573f3201a996c39e562f008303a2e1f69002931a6ad4ecb7bbce771dc4e4f: container already connected to network RED
```
### Hardware
- AWS
- All nodes are t2.micros
- All nodes are running stock Ubuntu 14.04 LTS
### Network

I have a 3 node setup on AWS with the following IPs:
- CIDR block: 10.10.0.0/16
- node1: 10.10.1.10 (consul master)
- node2: 10.10.2.10
- node3: 10.10.3.10
- All internal traffic on 10.10.0.0/16 is open.
### Software

On each box I executed the following to get docker installed.

```
# apt-get install linux-image-3.19.0-33-generic
# curl -sSL https://get.docker.com/ | sh
```
## Steps to reproduce:

```
node1:~# docker run -d -p 8500:8500 -h consul --name consul progrium/consul -server -bootstrap
node1:~# docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                                                            NAMES
d85135d0781e        progrium/consul     "/bin/start -server -"   3 hours ago         Up 16 minutes       53/tcp, 53/udp, 8300-8302/tcp, 8400/tcp, 8301-8302/udp, 0.0.0.0:8500->8500/tcp   consul

// Great! Consul is running on node1

node2:~# docker network create -d overlay --subnet=10.10.10.0/24 RED
node2:~# docker run -it --name container2 --rm --net RED busybox

// Dropped to a busybox shell on node2, container named **container2**

node3:~# docker run -it --name container3 --rm --net RED busybox
/ # ping container2
PING container2 (10.10.10.2): 56 data bytes
64 bytes from 10.10.10.2: seq=0 ttl=64 time=0.916 ms
64 bytes from 10.10.10.2: seq=1 ttl=64 time=0.816 ms
--- container2 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.816/0.866/0.916 ms

// I can also ping from container2 -> container3.  Everything looks good from here.
// Now I stop consul on node1 to simulate an outage or unreliable network link.
// I do this while both busybox shells are running on node2 and node3.

node1:~# docker stop consul

// back on node2, I exit out of busybox. 
// NOTE: This is with consul still stopped on node1!  This is important to reproduce
// After busybox on node2 has existed, I restart consul on node1

node1:~# docker start consul

// With consul back up I try to start a container on the RED network on node2

node2:~# docker run -it --name container2 --rm --net RED busybox
Error response from daemon: Cannot start container 1a99ec0339ca7116f47ea9c63a4a2868563a851b4f8bf40d18530a85d96f290a: container already connected to network RED

// I restart docker and try again but it complains about files existing already:

node2:~# service docker restart
docker stop/waiting
docker start/running, process 16893

# docker run -it --name container2 --rm --net RED busybox
Error response from daemon: Cannot start container 20b1998041019a5dc1e709a7814fd411d946a548f2f0644b1813d2f3c345c90a: subnet sandbox join failed for "10.10.10.0/24": error creating vxlan interface: file exists

// The only way I can get the container to start is to manually unmount and delete the network namespaces in /var/run/docker/netns e.g.

# service docker stop
docker stop/waiting
node2:~# umount /var/run/docker/netns/*
node2:~# rm /var/run/docker/netns/*
node2:~# service docker start
docker start/running, process 17138
node2:~# docker run -it --name container2 --rm --net RED busybox
/ # <--- busybox shell, success!
```
### node1 info

```
node1:~# docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

node1:~# uname -a
Linux ip-10-10-1-10 3.13.0-48-generic #80-Ubuntu SMP Thu Mar 12 11:16:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
### node2 info

```
node2:~# docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

node2:~# uname -a
Linux ip-10-10-2-10 3.19.0-37-generic #42~14.04.1-Ubuntu SMP Mon Nov 23 15:13:51 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

node2:~# cat /etc/default/docker 
DOCKER_OPTS="-H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock --cluster-store=consul://10.10.1.10:8500/network --cluster-advertise=eth0:2375"
```
### node3 info

```
node3:~# docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

node3:~# uname -a
Linux ip-10-10-3-10 3.19.0-33-generic #38~14.04.1-Ubuntu SMP Fri Nov 6 18:17:28 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

node3:~# cat /etc/default/docker 
DOCKER_OPTS="-H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock --cluster-store=consul://10.10.1.10:8500/network --cluster-advertise=eth0:2375"
```
## Reference:
- Steps mainly came from [here](https://github.com/docker/dceu_tutorials/blob/master/6-networking.md)

**Comment by mavenugo** [comment]
Thanks for the detailed reproduction steps @mgoldsborough.
We are planning on changing the logic of endpoint cleanup on such scenarios and that will solve the `container already connected to network RED` issue.

@sanimej was looking at the VXLAN device exists issue and I think the reproduction steps will help his investigation.

**Comment by mgoldsborough** [comment]
Great, thanks for the quick response @mavenugo!  Any idea how long it will take for this new logic to be implemented?  This is a major hangup in our multi-host networking efforts.  it would be great to use a native docker solution.

**Comment by mavenugo** [comment]
These are bug fixes and we will get them fixed shortly. But from a release standpoint, it will be available for Docker 1.10 (unless you are willing to try the docker master).

**Comment by mgoldsborough** [comment]
Okay, thanks.  I'd be happy to help from a testing standpoint but I'll probably wait until the official Docker 1.10 release to update my production servers :D
